# Predictions of Druggable peptides using the best model

In [58]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline

# remove warnings
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning) 

import numpy as np
import pandas as pd
import time
import matplotlib.pyplot as plt

from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_val_score, GridSearchCV, StratifiedKFold
from sklearn.metrics import confusion_matrix,accuracy_score, roc_auc_score,f1_score, recall_score, precision_score
from sklearn.utils import class_weight

from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression, LassoCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier
from xgboost import XGBClassifier
from sklearn.svm import SVC
from sklearn.gaussian_process.kernels import RBF
from sklearn.svm import LinearSVC

from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler
from sklearn.feature_selection import RFECV, VarianceThreshold, SelectKBest, chi2
from sklearn.feature_selection import SelectFromModel, SelectPercentile, f_classif

import seaborn as sns; sns.set() # data visualization library 
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import GradientBoostingClassifier, BaggingClassifier, AdaBoostClassifier
from sklearn.naive_bayes import BernoulliNB, GaussianNB
from imblearn.over_sampling import SMOTE

from sklearn.datasets import load_iris
from matplotlib import pyplot as plt
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, cross_val_score, KFold, StratifiedKFold
import numpy as np

import pandas as pd
from sklearn.utils import class_weight
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.externals import joblib

print(__doc__)

Automatically created module for IPython interactive environment


In [59]:
def DataCheckings(df):
    # CHECKINGS ***************************
    # Check the number of data points in the data set
    print("\nData points =", len(df))
    
    # Check the number of columns in the data set
    print("\nColumns (output + features)=",len(df.columns))
    
    # Check the data types
    print("\nData types =", df.dtypes.unique())
    
    # Dataset statistics
    print('\n')
    df.describe()
    
    # print names of columns
    print('Column Names:\n', df.columns)
    
    # see if there are categorical data
    print("\nCategorical features:", df.select_dtypes(include=['O']).columns.tolist())
    
    # Check NA values
    # Check any number of columns with NaN
    print("\nColumns with NaN: ", df.isnull().any().sum(), ' / ', len(df.columns))

    # Check any number of data points with NaN
    print("\nNo of data points with NaN:", df.isnull().any(axis=1).sum(), ' / ', len(df))

In [60]:
def getDataFromDataset(sFile, OutVar):
    # read details file
    print('\n-> Read dataset', sFile)
    df = pd.read_csv(sFile)
    #df = feather.read_dataframe(sFile)
    
    DataCheckings(df)
    
    # remove duplicates!
    df.drop_duplicates(keep=False, inplace=True)
    
    print('Shape', df.shape)
    # print(list(df.columns))

    # select X and Y
    ds_y = df[OutVar]
    ds_X = df.drop(OutVar,axis = 1)
    Xdata = ds_X.values # get values of features
    Ydata = ds_y.values # get output values

    print('Shape X data:', Xdata.shape)
    print('Shape Y data:',Ydata.shape)
    
    # return data for X and Y, feature names as list
    return (Xdata, Ydata, list(ds_X.columns))

In [61]:
def  set_weights(y_data, option='balanced'):
    """Estimate class weights for umbalanced dataset
       If ‘balanced’, class weights will be given by n_samples / (n_classes * np.bincount(y)). 
       If a dictionary is given, keys are classes and values are corresponding class weights. 
       If None is given, the class weights will be uniform """
    cw = class_weight.compute_class_weight(option, np.unique(y_data), y_data)
    w = {i:j for i,j in zip(np.unique(y_data), cw)}
    return w 

In [62]:
# define output variables
outVar = 'Class'

# define list of folds
foldType = 3

# define a label for output files
targetName = 'GS_Outer'

seed = 28

## Reproduce the pipeline model

In [63]:
sFile = './datasets/ds.Class_TC_ballanced.csv'

# get data from file
Xdata, Ydata, Features = getDataFromDataset(sFile,outVar) # n_sample=100


-> Read dataset ./datasets/ds.Class_TC_ballanced.csv

Data points = 1332

Columns (output + features)= 8001

Data types = [dtype('float64') dtype('int64')]


Column Names:
 Index(['AAA', 'RAA', 'NAA', 'DAA', 'CAA', 'EAA', 'QAA', 'GAA', 'HAA', 'IAA',
       ...
       'KVV', 'MVV', 'FVV', 'PVV', 'SVV', 'TVV', 'WVV', 'YVV', 'VVV', 'Class'],
      dtype='object', length=8001)

Categorical features: []

Columns with NaN:  0  /  8001

No of data points with NaN: 0  /  1332
Shape (1332, 8001)
Shape X data: (1332, 8000)
Shape Y data: (1332,)


In [64]:
# Calculate class weights
class_weights = set_weights(Ydata)
print("Class weights = ", class_weights)

Class weights =  {0: 1.0, 1: 1.0}


In [65]:
outer_cv = StratifiedKFold(n_splits=3,shuffle=True,random_state=seed)

In [9]:
ifold = 0
ACCs  =[]
AUROCs=[]
models =[]
SelectedFeatures =[]

for train_index, test_index in outer_cv.split(Xdata, Ydata):
    ifold +=1
    
    print("Fold =",ifold)
    start = time.time()
    
    #print("TRAIN:", train_index, "TEST:", test_index)
    X_train, X_test = Xdata[train_index], Xdata[test_index]
    y_train, y_test = Ydata[train_index], Ydata[test_index]
    
    # Standardize dataset
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test  = scaler.transform(X_test)
    
    # Feature selection # FS = SelectFromModel(LinearSVC(), max_features = 400,threshold=-np.inf)
    lsvc = LinearSVC(max_iter=50000).fit(X_train, y_train)
    model = SelectFromModel(lsvc, prefit=True,max_features = 200,threshold=-np.inf)
    X_train = model.transform(X_train)
    X_test  = model.transform(X_test)
    #print("Selected X:", X_train.shape)

    # Selected features
    SelFeatures = []
    for i in model.get_support(indices=True):
        SelFeatures.append(Features[i])
    SelectedFeatures.append(SelFeatures)

    #scaler.transform(X_test)
    clf = SVC(kernel = 'rbf', random_state=seed,gamma='scale',
              class_weight=class_weights,probability=True)
    clf.fit(X_train, y_train)
    
    joblib.dump(clf, 'SVM_model'+str(ifold)+'.pkl', compress = 1)
    models.append(clf)
    
    y_pred = clf.predict_proba(X_test)
    AUROC = roc_auc_score(y_test, y_pred[:, 1])
    AUROCs.append(AUROC)
    
    ACC = clf.score(X_test,y_test)
    ACCs.append(ACC)
   
    print("AUROC=",AUROC,"ACC=",ACC, (time.time() - start)/60,"mins")
    

Fold = 1
AUROC= 0.9707410112815518 ACC= 0.9301801801801802 2.053646918137868 mins
Fold = 2
AUROC= 0.9752049346643942 ACC= 0.9369369369369369 1.0309011777242025 mins
Fold = 3
AUROC= 0.9783499715932148 ACC= 0.9211711711711712 1.037073588371277 mins


Let's see the mean AUROC values for best model and the standard deviations:

In [10]:
print(np.mean(AUROCs),np.std(AUROCs))

0.974765305846387 0.0031218610239548486


In [11]:
print(np.mean(ACCs),np.std(ACCs))

0.9294294294294295 0.006458202152434386


In [12]:
# all the selected features for the 3 folds
print(SelectedFeatures)

[['NRA', 'LDA', 'KCA', 'EHA', 'THA', 'VIA', 'YLA', 'IMA', 'CSA', 'VYA', 'KNR', 'TDR', 'WDR', 'TER', 'PQR', 'WMR', 'PPR', 'RSR', 'VSR', 'ETR', 'FWR', 'NYR', 'PVR', 'NAN', 'YNN', 'YIN', 'FMN', 'HFN', 'HDD', 'SDD', 'ACD', 'MED', 'WED', 'TQD', 'SHD', 'WLD', 'VFD', 'WPD', 'SSD', 'LTD', 'PWD', 'EYD', 'VYD', 'WRC', 'TGC', 'IHC', 'WHC', 'VHC', 'WLC', 'GMC', 'GFC', 'GYC', 'EVC', 'GDE', 'SDE', 'SQE', 'IGE', 'NHE', 'NME', 'HME', 'YPE', 'TSE', 'AWE', 'EYE', 'QYE', 'SAQ', 'MRQ', 'INQ', 'PCQ', 'WEQ', 'AQQ', 'RQQ', 'KQQ', 'NGQ', 'CHQ', 'HLQ', 'HKQ', 'DSQ', 'FVQ', 'CDG', 'HGG', 'KGG', 'FGG', 'FPG', 'AWG', 'RVG', 'IVG', 'PAH', 'PDH', 'TQH', 'DGH', 'CGH', 'KHH', 'FWH', 'TYH', 'LVH', 'PAI', 'QNI', 'WEI', 'ESI', 'AYI', 'DRL', 'PDL', 'GPL', 'WYL', 'DVL', 'MVL', 'QRK', 'HNK', 'HCK', 'PHK', 'CIK', 'QLK', 'SMK', 'HTK', 'VYK', 'DVK', 'WRM', 'REM', 'FEM', 'WQM', 'DHM', 'SHM', 'LLM', 'TPM', 'KYM', 'IVM', 'WVM', 'YVM', 'CNF', 'LDF', 'SEF', 'EQF', 'YQF', 'VGF', 'YKF', 'AFF', 'CFF', 'FWF', 'CNP', 'QGP', 'PIP', 'AKP

In [13]:
# differences of selected descriptors
print(list(set(SelectedFeatures[1])-set(SelectedFeatures[0])))

['AVK', 'NKG', 'CHY', 'EKL', 'RWM', 'HGV', 'NFM', 'QSN', 'WHV', 'CHD', 'WSH', 'FHH', 'HVM', 'NDC', 'NSH', 'SSS', 'RSN', 'GYM', 'MLD', 'YAV', 'FNP', 'LWS', 'FEI', 'NPP', 'PSS', 'HGF', 'LIR', 'NDI', 'TTP', 'ERN', 'DHK', 'ESL', 'HDS', 'GNK', 'FDH', 'HKP', 'SMM', 'FYT', 'FNQ', 'HTD', 'CDH', 'YSQ', 'YGR', 'ALE', 'HDK', 'CLT', 'LPE', 'MCA', 'IKL', 'FFN', 'SSG', 'EWK', 'LHN', 'WNM', 'SQP', 'GEV', 'AWQ', 'DMY', 'FKL', 'INA', 'KLG', 'SHS', 'QGW', 'DFQ', 'YTT', 'QYV', 'DSV', 'IRT', 'FWN', 'FVE', 'RTG', 'GPQ', 'MGI', 'WQI', 'VYM', 'EQK', 'STP', 'KCW', 'LLV', 'VHP', 'MCE', 'HVW', 'NGF', 'TSM', 'SMD', 'WDS', 'LAT', 'WRI', 'SDN', 'FFK', 'QRA', 'TSV', 'GVE', 'RMQ', 'EHR', 'QSK', 'KKT', 'FAP', 'RVQ', 'VVL', 'WCD', 'TGG', 'ENV', 'QRG', 'IFH', 'YEA', 'DWD', 'EKK', 'MDN', 'PTG', 'PLP', 'MDQ', 'PEP', 'KVW', 'IKW', 'IMV', 'KNC', 'EDI', 'AAL', 'PMI']


In [14]:
# differences of selected descriptors
print(list(set(SelectedFeatures[1])-set(SelectedFeatures[2])))

['PQR', 'AVK', 'THA', 'CHY', 'RWM', 'NFM', 'IVG', 'QSN', 'WHV', 'CHD', 'WSH', 'FHH', 'HVM', 'NSH', 'QPP', 'FPG', 'SSS', 'GYM', 'MLD', 'LWS', 'FEI', 'NPP', 'PSS', 'HGF', 'NDI', 'WDR', 'DHK', 'ESL', 'HDS', 'GNK', 'VSR', 'HKP', 'SMM', 'FYT', 'FNQ', 'HTD', 'CDH', 'YSQ', 'YGR', 'ALE', 'HDK', 'TQH', 'CLT', 'KNR', 'LPE', 'MCA', 'IKL', 'FFN', 'SSG', 'EWK', 'LHN', 'SQP', 'VGW', 'AWQ', 'GEV', 'FKL', 'INA', 'QGW', 'DFQ', 'YTT', 'QYV', 'DSV', 'IRT', 'FWN', 'PCQ', 'ACD', 'FVE', 'MGI', 'WQI', 'VYM', 'EQK', 'STP', 'KCW', 'QGV', 'VHP', 'MCE', 'HVW', 'NGF', 'SMD', 'WDS', 'LAT', 'SDN', 'FFK', 'QRA', 'KYM', 'TSV', 'RMQ', 'QSK', 'FAP', 'FWH', 'WCD', 'TGG', 'CSA', 'QRG', 'IFH', 'EKK', 'AWE', 'PTG', 'MDQ', 'PEP', 'KVW', 'IMV', 'KNC', 'EDI', 'AAL', 'NHE', 'MED']


## Predictions with the best model

We choose model 2 as the best due to the maximum ACC value (AUROC= 0.9752, ACC= 0.937).

In [16]:
# the selected features for model 2
print(SelectedFeatures[1])

['NRA', 'QRA', 'INA', 'MCA', 'YEA', 'THA', 'CSA', 'VYA', 'KNR', 'WDR', 'TER', 'PQR', 'YGR', 'EHR', 'LIR', 'VSR', 'ERN', 'MDN', 'SDN', 'LHN', 'YIN', 'FFN', 'RSN', 'QSN', 'FWN', 'ACD', 'WCD', 'MED', 'CHD', 'SHD', 'MLD', 'SMD', 'WPD', 'SSD', 'HTD', 'DWD', 'VYD', 'KNC', 'NDC', 'IHC', 'VHC', 'GYC', 'MCE', 'NHE', 'ALE', 'HME', 'LPE', 'AWE', 'EYE', 'QYE', 'GVE', 'FVE', 'SAQ', 'FNQ', 'MDQ', 'PCQ', 'WEQ', 'RQQ', 'NGQ', 'HLQ', 'RMQ', 'DFQ', 'GPQ', 'DSQ', 'YSQ', 'AWQ', 'RVQ', 'QRG', 'HGG', 'TGG', 'KLG', 'NKG', 'FPG', 'SSG', 'RTG', 'PTG', 'IVG', 'CDH', 'FDH', 'PDH', 'TQH', 'KHH', 'FHH', 'IFH', 'NSH', 'WSH', 'FWH', 'WRI', 'NDI', 'EDI', 'FEI', 'WEI', 'WQI', 'MGI', 'PMI', 'AAL', 'EKL', 'IKL', 'FKL', 'GPL', 'ESL', 'DVL', 'MVL', 'VVL', 'GNK', 'HNK', 'HDK', 'HCK', 'EQK', 'DHK', 'QLK', 'EKK', 'SMK', 'FFK', 'QSK', 'EWK', 'AVK', 'WRM', 'WNM', 'REM', 'WQM', 'SHM', 'LLM', 'SMM', 'NFM', 'TSM', 'RWM', 'GYM', 'KYM', 'VYM', 'HVM', 'IVM', 'LDF', 'YQF', 'NGF', 'HGF', 'FWF', 'FAP', 'FNP', 'PEP', 'SQP', 'QGP', 'VHP'

Load the prediction datasets (the same format as the dataset: 8000 TC features + Class=-1):

In [68]:
# get data from files and check the files
sFile1 = './datasets/ds.Screening_1_TC.csv'
Xdata1, Ydata1, Features1 = getDataFromDataset(sFile1,outVar) 

sFile2 = './datasets/ds.Screening_2_TC.csv'
Xdata2, Ydata2, Features2 = getDataFromDataset(sFile2,outVar) 

sFile3 = './datasets/ds.Screening_3_TC.csv'
Xdata3, Ydata3, Features3 = getDataFromDataset(sFile3,outVar)

print(Xdata2.shape,Xdata3.shape)


-> Read dataset ./datasets/ds.Screening_1_TC.csv

Data points = 2359

Columns (output + features)= 8001

Data types = [dtype('float64') dtype('int64')]


Column Names:
 Index(['AAA', 'RAA', 'NAA', 'DAA', 'CAA', 'EAA', 'QAA', 'GAA', 'HAA', 'IAA',
       ...
       'KVV', 'MVV', 'FVV', 'PVV', 'SVV', 'TVV', 'WVV', 'YVV', 'VVV', 'Class'],
      dtype='object', length=8001)

Categorical features: []

Columns with NaN:  0  /  8001

No of data points with NaN: 0  /  2359
Shape (2353, 8001)
Shape X data: (2353, 8000)
Shape Y data: (2353,)

-> Read dataset ./datasets/ds.Screening_2_TC.csv

Data points = 231

Columns (output + features)= 8001

Data types = [dtype('float64') dtype('int64')]


Column Names:
 Index(['AAA', 'RAA', 'NAA', 'DAA', 'CAA', 'EAA', 'QAA', 'GAA', 'HAA', 'IAA',
       ...
       'KVV', 'MVV', 'FVV', 'PVV', 'SVV', 'TVV', 'WVV', 'YVV', 'VVV', 'Class'],
      dtype='object', length=8001)

Categorical features: []

Columns with NaN:  0  /  8001

No of data points with NaN: 0  /

Use only the second split / model - scale the prediction datasets, select only the features of model 2, predict the class and predict the probability of that class:

In [69]:
ifold = 0

for train_index, test_index in outer_cv.split(Xdata, Ydata):
    ifold +=1
    
    if ifold ==2: # only model 2
        print("Fold =",ifold)
        start = time.time()

        #print("TRAIN:", train_index, "TEST:", test_index)
        X_train, X_test = Xdata[train_index], Xdata[test_index]
        y_train, y_test = Ydata[train_index], Ydata[test_index]

        # Standardize dataset
        scaler = StandardScaler()
        X_train = scaler.fit_transform(X_train)
        X_test  = scaler.transform(X_test)
        # scale prediction set
        Xdata1  = scaler.transform(Xdata1)
        Xdata2  = scaler.transform(Xdata2)
        Xdata3  = scaler.transform(Xdata3)
        
        # Feature selection # FS = SelectFromModel(LinearSVC(), max_features = 400,threshold=-np.inf)
        lsvc = LinearSVC(max_iter=50000).fit(X_train, y_train)
        model = SelectFromModel(lsvc, prefit=True,max_features = 200,threshold=-np.inf)
        X_train = model.transform(X_train)
        X_test  = model.transform(X_test)
    
        # Selected features
        SelFeatures = []
        for i in model.get_support(indices=True):
            SelFeatures.append(Features[i])
        
        # apply selected features to prediction set
        Xdata1 = Xdata1[:, model.get_support()]
        print("Xdata1 sel=", Xdata1.shape)
        Xdata2 = Xdata2[:, model.get_support()]
        print("Xdata2 sel=", Xdata2.shape)
        Xdata3 = Xdata3[:, model.get_support()]
        print("Xdata3 sel=", Xdata3.shape)

        # we dont need to calculate again, but load the model from the disk!
        #clf = SVC(kernel = 'rbf', random_state=seed,gamma='scale',
        #          class_weight=class_weights,probability=True)
        #clf.fit(X_train, y_train)

        # load the saved model from disk
        clf = joblib.load('SVM_model'+str(ifold)+'.pkl')
        #joblib.dump(clf, 'SVM_model'+str(ifold)+'.pkl', compress = 1)
        
        # predictions with the model
        Ydata1 = clf.predict(Xdata1)
        Ydata2 = clf.predict(Xdata2)
        Ydata3 = clf.predict(Xdata3)
        
        # add probabilities (n_samples X n_classes; class 0, class 1)
        Ydata1prob = clf.predict_proba(Xdata1)
        Ydata2prob = clf.predict_proba(Xdata2)
        Ydata3prob = clf.predict_proba(Xdata3)
        
        # save predictions for list 1
        dff1 = pd.DataFrame(Xdata1,columns=SelFeatures)
        dff1['Class'] = Ydata1
        dff1['Prob0'] = Ydata1prob[:,0]
        dff1['Prob1'] = Ydata1prob[:,1]
        # merge with protein information from other file
        result = pd.concat([dff1, pd.read_csv('./PREDICTIONS/TC_seqs.Screening_1_Cancer_Driver_Genes.csv')], axis=1)
        # creat new order of columns in final results
        newHeader=['Class','Prob1','Prob0','V1','V2']
        result = result[newHeader]
        result = result.sort_values(by=['Prob1'], ascending=False)
        result.to_csv(sFile1[:-4]+'_predictions.csv', index=True)

        # save predictions for list 2
        dff2 = pd.DataFrame(Xdata2,columns=SelFeatures)
        dff2['Class'] = Ydata2
        dff2['Prob0'] = Ydata2prob[:,0]
        dff2['Prob1'] = Ydata2prob[:,1]
        # merge with protein information from other file
        result = pd.concat([dff2, pd.read_csv('./PREDICTIONS/TC_seqs.Screening_2_OncoOmics_Genes.csv')], axis=1)
        # creat new order of columns in final results
        newHeader=['Class','Prob1','Prob0','V1','V2']
        result = result[newHeader]
        result = result.sort_values(by=['Prob1'], ascending=False)
        result.to_csv(sFile2[:-4]+'_predictions.csv', index=True)
        
        # save predictions for list 3
        dff3 = pd.DataFrame(Xdata3,columns=SelFeatures)
        dff3['Class'] = Ydata3
        dff3['Prob0'] = Ydata3prob[:,0]
        dff3['Prob1'] = Ydata3prob[:,1]
        # merge with protein information from other file
        result = pd.concat([dff3, pd.read_csv('./PREDICTIONS/TC_seqs.Screening_3_RBPs.csv')], axis=1)
        # creat new order of columns in final results
        newHeader=['Class','Prob1','Prob0','V1','V2']
        result = result[newHeader]
        result = result.sort_values(by=['Prob1'], ascending=False)
        result.to_csv(sFile3[:-4]+'_predictions.csv', index=True)

        print("Time",(time.time() - start)/60,"mins")

Fold = 2
Xdata1 sel= (2353, 200)
Xdata2 sel= (231, 200)
Xdata3 sel= (1365, 200)
Time 1.0715562502543132 mins


In [70]:
print("==> Chekc the results:")
print(sFile1[:-4]+'_predictions.csv')
print(sFile2[:-4]+'_predictions.csv')
print(sFile3[:-4]+'_predictions.csv')

==> Chekc the results:
./datasets/ds.Screening_1_TC_predictions.csv
./datasets/ds.Screening_2_TC_predictions.csv
./datasets/ds.Screening_3_TC_predictions.csv


Hf with ML!@muntisa